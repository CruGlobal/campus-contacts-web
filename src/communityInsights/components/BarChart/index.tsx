import React, { ReactNode, useState } from 'react';
import { Bar } from '@nivo/bar';
import { withTheme } from 'emotion-theming';
import { useTranslation } from 'react-i18next';
import { line } from 'd3-shape';
import styled from '@emotion/styled';

import NullState from '../NullState';
import SwitchButton from '../SwitchButton';

const Wrapper = styled.div`
    height: 400px;
    position: relative;
`;

const Line = styled.div`
    border-top: 2px solid ${({ theme }) => theme.colors.highlight};
    width: 26px;
    height: 2px;

    &.dashed {
        border-top-style: dashed;
    }
`;

const Footer = styled.div`
    height: 32px;
    display: flex;
    justify-content: space-between;
`;

const Legend = styled.div`
    display: flex;
    align-items: center;
`;

const LegendLabel = styled.div`
    font-size: 12px;
    line-height: 16px;
    color: ${({ theme }) => theme.colors.secondary};
    margin-left: 5px;
    margin-right: 10px;
`;

const PreviousButton = styled.div`
    width: 6px;
    height: 12px;
    border-top: 6px solid transparent;
    border-bottom: 6px solid transparent;
    border-right: 6px solid #b4b6ba;
    position: absolute;
    bottom: 57px;
    left: 19px;
    cursor: pointer;

    &:hover {
        border-right: 6px solid #c3c3c3;
    }
`;

const NextButton = styled.div`
    width: 6px;
    height: 12px;
    border-top: 6px solid transparent;
    border-bottom: 6px solid transparent;
    border-left: 6px solid #b4b6ba;
    position: absolute;
    bottom: 57px;
    right: -10px;
    cursor: pointer;

    &:hover {
        border-left: 6px solid #c3c3c3;
    }
`;

const TooltipRow = styled.div`
    display: flex;
    justify-content: space-between;

    &.total {
        color: ${({ theme }) => theme.colors.highlight};
    }
`;

const Label = styled.div`
    font-weight: 600;
    font-size: 10px;
    line-height: 20px;
    text-transform: uppercase;
    letter-spacing: 1px;
    margin-right: 20px;
`;

const Value = styled.div`
    font-weight: 600;
    font-size: 10px;
    line-height: 20px;
    text-transform: uppercase;
    letter-spacing: 1px;
    align-self: flex-end;
`;

interface Props {
    theme: any;
    data: any;
    keys: any;
    indexBy: string;
    average?: number;
    filterType: string;
    title?: string;
    subtitle?: string;
    children?: ReactNode;
    index: number;
    datesFilter?: boolean;
    onFilterChanged: (type: string, index: number) => void;
    legendLabel?: string;
    tooltipBreakdown?: boolean;
    nullCheck: any;
    nullContent: string;
}

const BarChart = (props: Props) => {
    const {
        theme,
        data,
        keys,
        indexBy,
        average,
        tooltipBreakdown,
        onFilterChanged,
        index,
        filterType,
        datesFilter,
        legendLabel,
        nullCheck,
        nullContent,
    } = props;

    const { t } = useTranslation('insights');

    const [filter, setFilter] = useState({
        type: filterType,
        index,
    });

    const AverageLine = ({
        yScale,
        width,
    }: {
        yScale: (value: number) => number;
        width: number;
    }) => {
        if (!average) {
            return null;
        }

        const averageLine = [{ x: 0, y: average }, { x: width, y: average }];

        const lineGenerator = line<{ x: number; y: number }>()
            .x(({ x }) => x)
            .y(({ y }) => yScale(y));

        return (
            <path
                // @ts-ignore
                d={lineGenerator(averageLine)}
                fill="none"
                strokeDasharray="10, 5"
                strokeWidth="2"
                stroke="#B2ECF7"
            />
        );
    };

    const updateFilter = (index: number, type: string) => {
        setFilter({ type, index });
        onFilterChanged(type, index);
    };

    const onToggleFilterClick = (filterType: string) => {
        updateFilter(0, filterType);
    };

    const onNextRangeClick = () => {
        const newIndex = filter.index - 1 >= 0 ? filter.index - 1 : 0;
        updateFilter(newIndex, filter.type);
    };

    const onPreviousRangeClick = () => {
        const newIndex = filter.index + 1;
        updateFilter(newIndex, filter.type);
    };

    const tooltip = (options: any) => {
        const { data } = options;

        return [
            data.stages.map((row: { name: string; count: number }) => (
                <TooltipRow key={row.name}>
                    <Label>{row.name}</Label>
                    <Value>{row.count}</Value>
                </TooltipRow>
            )),
            <TooltipRow className={'total'} key={'total'}>
                <Label>Total</Label>
                <Value>{data.total}</Value>
            </TooltipRow>,
        ];
    };

    if (nullCheck) {
        return (
            <div>
                <NullState content={nullContent} />
                <Footer>
                    {datesFilter ? (
                        <SwitchButton
                            isMonth={filter.type === 'month'}
                            leftLabel={t('monthLabel')}
                            rightLabel={t('yearLabel')}
                            onLeftClick={() => onToggleFilterClick('month')}
                            onRightClick={() => onToggleFilterClick('year')}
                        />
                    ) : null}
                    {legendLabel ? (
                        <Legend>
                            <Line />
                            <LegendLabel>{legendLabel}</LegendLabel>
                            <Line className={'dashed'} />
                            <LegendLabel>{t('average')}</LegendLabel>
                        </Legend>
                    ) : null}
                </Footer>
            </div>
        );
    }

    return (
        <div>
            <Wrapper>
                <Bar
                    height={400}
                    width={848}
                    data={data}
                    keys={keys}
                    indexBy={indexBy}
                    theme={theme.graph}
                    // @ts-ignore
                    layers={[
                        AverageLine,
                        'grid',
                        'axes',
                        'bars',
                        'markers',
                        'legends',
                    ]}
                    margin={{
                        top: 30,
                        right: 0,
                        bottom: 90,
                        left: 30,
                    }}
                    minValue={'auto'}
                    maxValue={'auto'}
                    padding={0.4}
                    innerPadding={3}
                    colors={[
                        theme.colors.highlight,
                        theme.colors.highlightDarker,
                    ]}
                    colorBy="id"
                    axisTop={null}
                    groupMode={'grouped'}
                    axisRight={null}
                    // @ts-ignore
                    tooltip={tooltipBreakdown ? tooltip : null}
                    axisBottom={{
                        tickSize: 0,
                        tickPadding: 20,
                    }}
                    axisLeft={{
                        tickSize: 0,
                        tickPadding: 5,
                        tickRotation: 0,
                    }}
                    enableGridY={true}
                    enableLabel={false}
                    animate={true}
                    motionStiffness={90}
                    motionDamping={15}
                />
                {datesFilter ? (
                    <div>
                        <PreviousButton onClick={onPreviousRangeClick} />
                        <NextButton onClick={onNextRangeClick} />
                    </div>
                ) : null}
            </Wrapper>
            <Footer>
                {datesFilter ? (
                    <SwitchButton
                        isMonth={filter.type === 'month'}
                        leftLabel={t('monthLabel')}
                        rightLabel={t('yearLabel')}
                        onLeftClick={() => onToggleFilterClick('month')}
                        onRightClick={() => onToggleFilterClick('year')}
                    />
                ) : null}
                {legendLabel ? (
                    <Legend>
                        <Line />
                        <LegendLabel>{legendLabel}</LegendLabel>
                        <Line className={'dashed'} />
                        <LegendLabel>{t('average')}</LegendLabel>
                    </Legend>
                ) : null}
            </Footer>
        </div>
    );
};

export default withTheme(BarChart);
