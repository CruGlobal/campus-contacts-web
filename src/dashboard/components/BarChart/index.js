import React from 'react';
import PropTypes from 'prop-types';
import { ResponsiveBar } from '@nivo/bar';
import { withTheme } from 'emotion-theming';
import { line } from 'd3-shape';

const BarChart = props => {
    const { theme, data, keys, indexBy, average, tooltip } = props;

    const AverageLine = ({ bars, xScale, yScale }) => {
        if (!average) {
            return null;
        }

        const lineGenerator = line()
            .x(d => {
                const val = xScale(d.data.index) + 2 * d.width + 50;
                if (!val) {
                    return 0;
                }
                return val;
            })
            .y(d => yScale(average));

        return (
            <path
                d={lineGenerator(bars)}
                fill="none"
                strokeDasharray="10, 5"
                strokeWidth="2"
                stroke="#B2ECF7"
            />
        );
    };

    return (
        <ResponsiveBar
            data={data}
            keys={keys}
            indexBy={indexBy}
            theme={theme.graph}
            layers={[AverageLine, 'grid', 'axes', 'bars', 'markers', 'legends']}
            margin={{
                top: 30,
                right: 0,
                bottom: 90,
                left: 30,
            }}
            minValue={'auto'}
            maxValue={'auto'}
            padding={0.4}
            innerPadding={3}
            colors={[theme.colors.highlight, theme.colors.highlightDarker]}
            colorBy="id"
            axisTop={null}
            groupMode={'grouped'}
            axisRight={null}
            tooltip={tooltip}
            axisBottom={{
                tickSize: 0,
                tickPadding: 20,
            }}
            axisLeft={{
                tickSize: 0,
                tickPadding: 5,
                tickRotation: 0,
            }}
            enableGridY={true}
            enableLabel={false}
            animate={true}
            motionStiffness={90}
            motionDamping={15}
        />
    );
};

export default withTheme(BarChart);

BarChart.propTypes = {
    title: PropTypes.string,
    subtitle: PropTypes.string,
    children: PropTypes.element,
};
