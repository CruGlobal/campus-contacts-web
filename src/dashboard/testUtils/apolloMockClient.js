import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
import { buildClientSchema } from 'graphql/utilities';
import { createMockerFromIntrospection } from 'fraql/mock';
import gql from 'fraql';

import introspectionQuery from '../../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument, options) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
