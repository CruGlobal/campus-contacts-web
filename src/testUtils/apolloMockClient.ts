import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema } from 'graphql-tools';
// @ts-ignore
import { buildClientSchema } from 'graphql/utilities';
// @ts-ignore
import { createMockerFromIntrospection } from 'fraql/mock';
// @ts-ignore
import gql from 'fraql';

import introspectionQuery from '../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mocks = {}) => {
    const schema = buildClientSchema(introspectionQuery);

    addMockFunctionsToSchema({
        schema,
        mocks: { ...globalMocks, ...mocks },
    });

    return new ApolloClient({
        cache: new InMemoryCache(),
        link: new SchemaLink({ schema }),
    });
};

export const mockFragment = (fragmentDocument: any, options: any) => {
    return createMockerFromIntrospection(introspectionQuery, {
        mocks: globalMocks,
    }).mockFragment(gql(fragmentDocument.loc.source.body), options);
};
