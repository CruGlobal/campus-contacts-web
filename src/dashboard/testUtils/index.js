import React, { ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';

import { createApolloMockClient } from './apolloMockClient';
import defaultTheme from '../defaultTheme';
import { ThemeProvider } from 'emotion-theming';
import AppContext from '../appContext';

export const renderWithContext = (component, { mocks } = {}, context = {}) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={context}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });

    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
