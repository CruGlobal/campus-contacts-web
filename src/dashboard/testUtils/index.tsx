import React, { ReactNode } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';

import defaultTheme from '../defaultTheme';
import AppContext from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

export const renderWithContext = (
    component: any,
    { mocks: mocks = {}, appContext = {} } = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    // @ts-ignore
    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
