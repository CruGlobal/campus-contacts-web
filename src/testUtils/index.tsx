import React, { ReactNode, ReactElement } from 'react';
import { ApolloProvider } from 'react-apollo-hooks';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import { IMocks } from 'graphql-tools';

import defaultTheme from '../defaultTheme';
import {
    AppContext,
    AppContextValue,
    appContextDefaultValue,
} from '../appContext';

import { createApolloMockClient } from './apolloMockClient';

interface RenderWithContextParams {
    mocks?: IMocks;
    appContext?: AppContextValue;
}

export const renderWithContext = (
    component: ReactElement,
    {
        mocks: mocks = {},
        appContext = appContextDefaultValue,
    }: RenderWithContextParams = {},
) => {
    const mockApolloClient = createApolloMockClient(mocks);

    // Warning: don't call any functions in here that return new instances on every call.
    // All the props need to stay the same otherwise renderer won't work.
    const wrapper = ({ children }: { children?: ReactNode }) => (
        <ApolloProvider client={mockApolloClient}>
            <AppContext.Provider value={appContext}>
                <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
            </AppContext.Provider>
        </ApolloProvider>
    );

    const renderResult = render(component, { wrapper });
    return {
        ...renderResult,
        snapshot: () => {
            expect(renderResult.container).toMatchSnapshot();
        },
    };
};
