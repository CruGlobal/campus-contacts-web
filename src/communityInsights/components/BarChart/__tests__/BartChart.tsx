import React from 'react';

import { renderWithContext } from '../../../../testUtils';

import BarChart from '..';

describe('<BarChart />', () => {
    it('should render properly', async () => {
        renderWithContext(
            <BarChart
                nullContent={'stepsOfFaithCompleted'}
                data={[]}
                keys={['a']}
                indexBy={'b'}
                filterType={'month'}
                onFilterChanged={() => {}}
                index={0}
            />,
        ).snapshot();
    });

    it('should render properly with data, key, index by', async () => {
        const graphData = [{ a: 1, b: 2 }, { a: 2, b: 3 }, { a: 3, b: 4 }];

        renderWithContext(
            <BarChart
                nullContent={'stepsOfFaithCompleted'}
                data={graphData}
                keys={['a']}
                indexBy={'b'}
                filterType={'month'}
                onFilterChanged={() => {}}
                index={0}
            />,
        ).snapshot();
    });
});
