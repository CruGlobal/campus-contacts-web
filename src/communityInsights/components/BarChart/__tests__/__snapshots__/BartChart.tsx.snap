// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BarChart /> should render properly 1`] = `
.emotion-1 {
  width: 100%;
  height: 190px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #ffffff;
  border-radius: 8px;
}

.emotion-0 {
  text-align: center;
  width: 720px;
  color: #007398;
  font-size: 16px;
  font-weight: 300;
  line-height: 24px;
  margin: 10px 0;
}

.emotion-2 {
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

<div>
  <div>
    <div
      class="emotion-1"
    >
      <img
        src="mockFile"
      />
      <p
        class="emotion-0"
      >
        This community does not have any members who have completed a personal step of faith. Challenge your community to open the MissionHub mobile app and complete a personal step of faith.
      </p>
    </div>
    <div
      class="emotion-2"
    />
  </div>
</div>
`;

exports[`<BarChart /> should render properly with data, key, index by 1`] = `
.emotion-1 {
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-0 {
  height: 400px;
  position: relative;
}

<div>
  <div>
    <div
      class="emotion-0"
    >
      <div
        style="position: relative;"
      >
        <svg
          height="400"
          role="img"
          width="848"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            fill="transparent"
            height="400"
            width="848"
          />
          <g
            transform="translate(30,30)"
          >
            <g>
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="280"
                y2="280"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="261"
                y2="261"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="243"
                y2="243"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="224"
                y2="224"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="205"
                y2="205"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="187"
                y2="187"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="168"
                y2="168"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="149"
                y2="149"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="131"
                y2="131"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="112"
                y2="112"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="93"
                y2="93"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="75"
                y2="75"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="56"
                y2="56"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="37"
                y2="37"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="19"
                y2="19"
              />
              <line
                opacity="1"
                stroke="#ECEEF2"
                stroke-width="1"
                x1="0"
                x2="818"
                y1="0"
                y2="0"
              />
            </g>
            <g
              transform="translate(0,280)"
            >
              <g
                style="opacity: 1;"
                transform="translate(169,0)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="text-before-edge"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="middle"
                  transform="translate(0,20) rotate(0)"
                >
                  2
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(409,0)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="text-before-edge"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="middle"
                  transform="translate(0,20) rotate(0)"
                >
                  3
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(649,0)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="text-before-edge"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="middle"
                  transform="translate(0,20) rotate(0)"
                >
                  4
                </text>
              </g>
              <line
                style="stroke: transparent; stroke-width: 1;"
                x1="0"
                x2="818"
                y1="0"
                y2="0"
              />
            </g>
            <g
              transform="translate(0,0)"
            >
              <g
                style="opacity: 1;"
                transform="translate(0,280)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  0
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,261)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  0.2
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,243)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  0.4
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,224)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  0.6
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,205)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  0.8
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,187)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  1
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,168)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  1.2
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,149)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  1.4
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,131)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  1.6
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,112)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  1.8
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,93)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  2
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,75)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  2.2
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,56)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  2.4
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,37)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  2.6
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,19)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  2.8
                </text>
              </g>
              <g
                style="opacity: 1;"
                transform="translate(0,0)"
              >
                <line
                  style="stroke: #777777; stroke-width: 1;"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="0"
                />
                <text
                  dominant-baseline="central"
                  style="font-weight: 600; letter-spacing: 1px; font-family: Source Sans Pro; font-size: 10px; fill: #B4B6BA;"
                  text-anchor="end"
                  transform="translate(-5,0) rotate(0)"
                >
                  3
                </text>
              </g>
              <line
                style="stroke: transparent; stroke-width: 1;"
                x1="0"
                x2="0"
                y1="0"
                y2="280"
              />
            </g>
            <g>
              <g
                transform="translate(97, 187)"
              >
                <rect
                  fill="#3CC8E6"
                  height="93"
                  rx="0"
                  ry="0"
                  stroke="#3CC8E6"
                  stroke-width="0"
                  width="144"
                />
              </g>
              <g
                transform="translate(337, 93)"
              >
                <rect
                  fill="#3CC8E6"
                  height="187"
                  rx="0"
                  ry="0"
                  stroke="#3CC8E6"
                  stroke-width="0"
                  width="144"
                />
              </g>
              <g
                transform="translate(577, 0)"
              >
                <rect
                  fill="#3CC8E6"
                  height="280"
                  rx="0"
                  ry="0"
                  stroke="#3CC8E6"
                  stroke-width="0"
                  width="144"
                />
              </g>
            </g>
          </g>
        </svg>
      </div>
    </div>
    <div
      class="emotion-1"
    />
  </div>
</div>
`;
